ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore/Mock.hs

```diff
@@ -0,0 +1,338 @@
+  | ErrBackingStoreValueHandleClosed
+  | ErrCopyPathAlreadyExists
+  | ErrCopyPathDoesNotExist
+  | ErrNonMonotonicSeqNo (WithOrigin SlotNo) (WithOrigin SlotNo)
+  deriving stock (Show, Eq)
+
+{-------------------------------------------------------------------------------
+  Type classes
+-------------------------------------------------------------------------------}
+
+-- | Abstract over interactions between values, keys and diffs.
+class ( EmptyValues vs, ApplyDiff vs d, LookupKeysRange ks vs
+      , LookupKeys ks vs, ValuesLength vs, MakeDiff vs d
+      , DiffSize d, KeysSize ks
+      ) => HasOps ks vs d
+
+class EmptyValues vs where
+  emptyValues :: vs
+
+class ApplyDiff vs d where
+  applyDiff :: vs -> d -> vs
+
+class LookupKeysRange ks vs where
+  lookupKeysRange :: Maybe ks -> Int -> vs -> vs
+
+class LookupKeys ks vs where
+  lookupKeys :: ks -> vs -> vs
+
+class ValuesLength vs where
+  valuesLength :: vs -> Int
+
+class MakeDiff vs d where
+  diff :: vs -> vs -> d
+
+class DiffSize d where
+  diffSize :: d -> Int
+
+class KeysSize ks where
+  keysSize :: ks -> Int
+
+{-------------------------------------------------------------------------------
+  State monad to run the mock in
+-------------------------------------------------------------------------------}
+
+-- | State within which the mock runs.
+newtype MockState ks vs d a =
+    MockState (ExceptT Err (State (Mock vs)) a)
+  deriving stock     Functor
+  deriving newtype ( Applicative
+                   , Monad
+                   , MonadState (Mock vs)
+                   , MonadError Err
+                   )
+
+runMockState ::
+     MockState ks vs d a
+  -> Mock vs
+  -> (Either Err a, Mock vs)
+runMockState (MockState t) = runState . runExceptT $ t
+
+{------------------------------------------------------------------------------
+  Mocked @'BackingStore'@ operations
+------------------------------------------------------------------------------}
+
+mBSInitFromValues ::
+     forall vs m. (MonadState (Mock vs) m)
+  => WithOrigin SlotNo
+  -> vs
+  -> m ()
+mBSInitFromValues sl vs = modify (\m -> m {
+    backingValues = vs
+  , backingSeqNo  = sl
+  , isClosed      = False
+  })
+
+mBSInitFromCopy ::
+     forall vs m. (MonadState (Mock vs) m, MonadError Err m)
+  => FS.FsPath
+  -> m ()
+mBSInitFromCopy bsp = do
+  cps <- gets copies
+  case Map.lookup bsp cps of
+    Nothing       -> throwError ErrCopyPathDoesNotExist
+    Just (sl, vs) -> modify (\m -> m {
+        backingValues = vs
+      , backingSeqNo  = sl
+  , isClosed      = False
+      })
+
+-- | Throw an error if the backing store has been closed.
+mGuardBSClosed :: (MonadState (Mock vs) m, MonadError Err m) => m ()
+mGuardBSClosed = do
+  closed <- gets isClosed
+  when closed $
+    throwError ErrBackingStoreClosed
+
+-- | Close the backing store.
+--
+-- Closing is idempotent.
+mBSClose :: (MonadState (Mock vs) m, MonadError Err m) => m ()
+mBSClose = (mGuardBSClosed >> close) `catchError` handler
```

Comment:
Just branch on `gets isClosed` instead of this exception handler hoop jumping.

Reply:
Done